src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/HandshakeEffect.kt

```diff
@@ -0,0 +1,42 @@
+package com.pubnub.api.subscribe.eventengine.effect
+
+import com.pubnub.api.PubNubException
+import com.pubnub.api.endpoints.remoteaction.RemoteAction
+import com.pubnub.api.eventengine.EventDeliver
+import com.pubnub.api.eventengine.ManagedEffect
+import com.pubnub.api.subscribe.eventengine.event.Event
+import com.pubnub.api.subscribe.eventengine.event.SubscriptionCursor
+
+class HandshakeEffect(
+    private val handshakeProvider: HandshakeProvider,
+    private val eventDeliver: EventDeliver,
+    private val handshakeInvocation: SubscribeEffectInvocation.Handshake
+) : ManagedEffect {
+    override fun runEffect(completionBlock: () -> Unit) {
+        val handshakeRemoteAction = getHandshakeRemoteAction()
+        handshakeRemoteAction.async { result, status ->
+            try {
+                if (status.error) {
+                    eventDeliver.passEventForHandling(
+                        Event.HandshakeFailure(
+                            status.exception
+                                ?: PubNubException("Unknown error") // todo check it that can happen
+                        )
+                    )
+                } else {
+                    eventDeliver.passEventForHandling(Event.HandshakeSuccess(result!!))
+                }
+            } finally {
+                completionBlock()
+            }
+        }
+    }
+
+    override fun cancel() {
+        val handshakeRemoteAction = getHandshakeRemoteAction()
```

Comment:
This will only work if `getHandshakeRemoteAction` returns same instance. I don't think we should assume this therefore this is a bug. It's not possible to really cancel this effect. Same mistake is present in other classes.

Reply:
Make sense. I will modify it.